_posts/2024-11-15-WildFly-playing-with-generative-ai.adoc

```diff
@@ -0,0 +1,347 @@
+                <goal>package</goal>
+            </goals>
+        </execution>
+    </executions>
+</plugin>
+----
+
+==== The code
+
+The code itself is quite straightforward:
+[source, java]
+----
+@ServerEndpoint(value = "/websocket/chatbot",
+        configurator = org.wildfly.ai.websocket.CustomConfigurator.class)
+public class RagChatBot {
+    @Inject
+    @Named(value = "ollama")
+    ChatLanguageModel chatModel;
+    @Inject
+    @Named(value = "embedding-store-retriever")
+    ContentRetriever retriever;
+
+  private static final String PROMPT_TEMPLATE = "You are a WildFly expert who understands well how to administrate the WildFly server and its components\n"
+            + "Objective: answer the user question delimited by  ---\n"
+            + "\n"
+            + "---\n"
+            + "{{userMessage}}\n"
+            + "---"
+            + "\n Here is a few data to help you:\n"
+            + "{{contents}}";
+
+  @OnMessage
+  public String sayHello(String question, Session session) throws IOException {
+        ChatMemory chatMemory = MessageWindowChatMemory.builder().id(session.getUserProperties().get("httpSessionId")).maxMessages(3).build();
+        ConversationalRetrievalChain chain = ConversationalRetrievalChain.builder()
+                .chatLanguageModel(chatModel)
+                .chatMemory(chatMemory)
+                .retrievalAugmentor(createBasicRag())
+                .build();
+        String result = chain.execute(question).replace("\n", "<br/>");
+        return result;
+    }
+
+    private RetrievalAugmentor createBasicRag() {
+        return DefaultRetrievalAugmentor.builder()
+                .contentRetriever(retriever)
+                .contentInjector(DefaultContentInjector.builder()
+                        .promptTemplate(PromptTemplate.from(PROMPT_TEMPLATE))
+                        .build())
+                .queryRouter(new DefaultQueryRouter(retriever))
+                .build();
+    }
+}
+----
+
+The `dev.langchain4j.rag.content.retriever.ContentRetriever` called *embedding-store-retriever* defined in the subsystem is injected into our WebSocket endpoint using the the `@Named` annotation.
+
+It is used to instantiate a `dev.langchain4j.rag.RetrievalAugmentor` which is in charge with retrieving the contents and enriching the prompt with them.
+
+In the same way, the `dev.langchain4j.model.chat.ChatLanguageModel` called *ollama* defined in the subsystem is injected into our WebSocket endpoint using the `@Named` annotation.
+
+With those two elements, a `dev.langchain4j.chain.ConversationalRetrievalChain` is created and used to interact with the LLM and send back the answer to the client using the WebSocket.
+
+The final code of the application is a bit more complex than what is exposed here as it tries to keep some context between a user queries.
+
+==== Building and Running the application
+
+First you need to clone it from https://github.com/ehsavoie/webchat[https://github.com/ehsavoie/webchat].
```

Comment:
general tip for all links in this file:

https://wildfly.zulipchat.com/#narrow/channel/174184-wildfly-developers/topic/Tip.20for.20wildfly.2Eorg.20content